zetaclient/types/event.go

```diff
@@ -41,3 +65,47 @@ type InboundEvent struct {
 	// Asset is the asset of the inbound
 	Asset string
 }
+
+// DecodeMemo decodes the receiver from the memo bytes
+func (event *InboundEvent) DecodeMemo() error {
```

Comment:
All of this logic should be implemented within `NewInboundEvent(...)`. The struct itself should be immutable.

Regarding this package per se: this is what we currently have for SOL:
`solrpc.GetTransactionResult` → `solana.Transaction` → `solanacontracts.Deposit` → `clienttypes.InboundEvent` → `crosschain.MsgVoteInbound`. 

I don't understand why it can't be `solana.Transaction` → `solanacontracts.Deposit` → `crosschain.MsgVoteInbound`

Reply:
It is a good point. This event struct was intended as a general struct for non-EVM chains [here](https://github.com/zeta-chain/node/issues/2495).

Now, it seems that Solana is good enough to just use its own instruction structs (Deposit, DepositSPL), we could move the logic of Solana memo to its own structs, leaving it as a general struct for UTXO chains (e.g., Doge): `transaction` -> `InboundEvent` -> `MsgVoteInbound`